scripts/usb_scan.py

```diff
@@ -0,0 +1,748 @@
+        # usb_intf_attr(bInterfaceProtocol, "%02x\n");
+        for p in self._PROPS:
+            if p not in self or not self.validate_int(self[p], 16):
+                return False
+        return super(UsbInterface, self).is_initialized()
+
+    def is_child_of(self, parent):
+        if type(parent) is UsbDevice and parent.is_initialized():
+            conf_value = parent[UsbDevice._CONF_VALUE]
+            pattern = r"^{}:{}\.\d+$".format(re.escape(parent.get_node()), re.escape(conf_value))
+            return re.match(pattern, self.get_node()) is not None
+        return False
+
+    def get_storages(self):
+        """ getter of UsbStorage list for mass storage interface
+
+        :return:
+        """
+        return self.storages
+
+    def add_storage(self, storage):
+        """ add a storage to this interface
+
+        :param storage(UsbStorage): the storage to add
+        :return: None
+        """
+        if storage in self.storages:
+            Log().debug("overriding existing storage: " + storage)
+            self.storages.remove(storage)
+        self.storages.add(storage)
+
+    def del_storage(self, storage):
+        """ remove a storage to this interface
+
+        :param storage:(UsbStorage) the storage to remove
+        :return: None
+        """
+        if storage in self.storages:
+            self.storages.remove(storage)
+
+
+class UsbStorage(UsbObject):
+    """ Class for USB block disk, save full device path property in UsbObject dict
+
+    """
+    _PATH = "path"
+    _PROPS = [_PATH]
+
+    def __init__(self, node, path):
+        """ initialise UsbStorage, set node and properties
+
+        :param node:(str) device node
+        :param path:(str) full device path
+        """
+        super(UsbStorage, self).__init__(node)
+        self[self._PATH] = path
+
+    def is_class_hub(self):
+        return False
+
+    def is_initialized(self):
+        path = self.get(self._PATH)
+        if path is None or path.find("usb") == -1:
+            return False
+        return super(UsbStorage, self).is_initialized()
+
+    def get_parent_node(self):
+        """get node of parent UsbInterface
+
+        :return: (str) parent node
+        """
+        parts = self[self._PATH].split("/")
+        usb = False
+        for part in parts:
+            if usb:
+                if part.find(":") != -1:
+                    return part
+            elif part.startswith("usb"):
+                usb = True
+        return None
+
+    def is_child_of(self, parent):
+        if type(parent) is UsbInterface and parent.is_initialized():
+            return parent.get_node() == self.get_parent_node()
+        return False
+
+
+class Singleton(type):
+    _instances = {}
+
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
+        return cls._instances[cls]
+
+
+class Log(object):
```

Comment:
Do we really need to roll out our own Singleton and Logger when we could just use the xapi one instead?

Reply:
Because I want to show log in terminal directly when debugging,
This Log class allow me to enable printing easily.

Reply:
I think this should not be part of the script in xapi, adds unnecessary complexity and on the host it has to log with the other scripts. 

You could redefine the logger on the top of the file while testing it

Reply:
OK, to remove the Log class.